### PR TITLE
adapted AutoUpdateDelegate to the new ESH auto update infrastructure

### PR DESCRIPTION
**Pull request by kaikreuzer**
Related to https://github.com/eclipse/smarthome/pull/5011

While at it, I also changed the DS components to use annotations instead of XMLs.

Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Note to reviewers: This code will only compile / can only be merged once https://github.com/eclipse/smarthome/pull/5011 is merged and a new ESH stable has been built.

**Comment by kaikreuzer** [comment]
FTR, https://github.com/eclipse/smarthome/pull/5011 is merged. I'd do a new ESH Stable once one of @openhab/2-x-add-ons-maintainers approves this PR.

File: bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java [hunk 1]
```diff
@@ -20,26 +37,185 @@
...
+        this.listeners.remove(listener);
+    }
+
+    private void refreshItemUpdateVetos() {
+        Set<String> newVetos = new HashSet<>();
+        synchronized (itemUpdateVetos) {
+            itemUpdateVetos.clear();
+            for (Item item : itemRegistry.getAll()) {
+                for (org.openhab.core.autoupdate.AutoUpdateBindingProvider provider : providers) {
+                    Boolean autoUpdate = provider.autoUpdate(item.getName());
```
Comment (martinvw) [inline bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java]: Does it really return a `Boolean` with potential null's?

File: bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java [hunk 2]
```diff
@@ -20,26 +37,185 @@
...
+    public void added(Item element) {
+        String itemName = element.getName();
+        refreshVetoForItem(itemName);
+    }
+
+    private void refreshVetoForItem(String itemName) {
+        synchronized (itemUpdateVetos) {
+            boolean removed = itemUpdateVetos.remove(itemName);
+            for (org.openhab.core.autoupdate.AutoUpdateBindingProvider provider : providers) {
+                Object autoUpdate = provider.autoUpdate(itemName);
```
Comment (martinvw) [inline bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java]: In other places we at least use a `Boolean`...

File: bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java [hunk 3]
```diff
@@ -20,26 +37,185 @@
...
+    protected void deactivate() {
+        for (org.openhab.core.autoupdate.AutoUpdateBindingProvider provider : providers) {
+            provider.removeBindingChangeListener(this);
+        }
+        itemRegistry.removeRegistryChangeListener(this);
+        started = false;
+    }
 
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE)
     public void addAutoUpdateBindingProvider(org.openhab.core.autoupdate.AutoUpdateBindingProvider provider) {
```
Comment (martinvw) [inline bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java]: It seems we can import these now, given the one from ESH is not imported anymore.

File: bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java [hunk 4]
```diff
@@ -20,26 +37,185 @@
...
+    public void removeProviderChangeListener(ProviderChangeListener<Metadata> listener) {
+        this.listeners.remove(listener);
+    }
+
+    private void refreshItemUpdateVetos() {
+        Set<String> newVetos = new HashSet<>();
+        synchronized (itemUpdateVetos) {
+            itemUpdateVetos.clear();
+            for (Item item : itemRegistry.getAll()) {
+                for (org.openhab.core.autoupdate.AutoUpdateBindingProvider provider : providers) {
```
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java]: Yes, indeed, now that it is removed from ESH :-)

File: bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java]: Yes. A provider returns null if it does not want to do any statement.

File: bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java (hunk 2 above)
Comment (kaikreuzer) [reply bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/autoupdate/internal/AutoUpdateProviderDelegate.java]: true

**Comment by kaikreuzer** [comment]
Thanks - I have updated the PR.

**Comment by kaikreuzer** [comment]
Retriggering PR builds...

**Comment by kaikreuzer** [comment]
Nice, DS annotations actually caught a bug in the unset method. Fixed it with https://github.com/openhab/openhab-core/pull/390/commits/8f3dea0a448fcf53179c03e1a2ef08c9bd71d548.

**Comment by kaikreuzer** [comment]
@martinvw All green now, so I think we can merge!

**Pull request merged**